Wrappers/Python/ccpi/optimisation/functions/Function.py

```diff
@@ -23,55 +23,493 @@
+                            
+            
+        
+class ScaledFunction(Function):
+    
+    r""" ScaledFunction represents the scalar multiplication with a Function.
+    
+    Let a function F then and a scalar :math:`\alpha`.
+    
+    If :math:`G(x) = \alpha F(x)` then:
+    
+    1. :math:`G(x) = \alpha  F(x)` ( __call__ method )
+    2. :math:`G'(x) = \alpha  F'(x)` ( gradient method ) 
+    3. :math:`G^{*}(x^{*}) = \alpha  F^{*}(\frac{x^{*}}{\alpha})` ( convex_conjugate method )   
+    4. :math:`\mathrm{prox}_{\tau G}(x) = \mathrm{prox}_{(\tau\alpha) F}(x)` ( proximal method ) 
+           
+    """
+    def __init__(self, function, scalar):
+        
+        super(ScaledFunction, self).__init__() 
+                                                     
+        if not isinstance (scalar, Number):
+            raise TypeError('expected scalar: got {}'.format(type(scalar)))
+        
+        if function.L is not None:        
+            self.L = abs(scalar) * function.L  
+            
+        self.scalar = scalar
+        self.function = function       
+              
+    def __call__(self,x, out=None):
+        r"""Returns the value of the scaled function.
+        
+        .. math:: G(x) = \alpha F(x)
+        
+        """
+        return self.scalar * self.function(x)
+
+    def convex_conjugate(self, x):
+        r"""Returns the convex conjugate of the scaled function.
+        
+        .. math:: G^{*}(x^{*}) = \alpha  F^{*}(\frac{x^{*}}{\alpha})
+        
+        """
+        return self.scalar * self.function.convex_conjugate(x/self.scalar)
+    
+    def gradient(self, x, out=None):
+        r"""Returns the gradient of the scaled function.
+        
+        .. math:: G'(x) = \alpha  F'(x)
+        
+        """
+        
+#        try:
+        if out is None:            
+            return self.scalar * self.function.gradient(x)
+        else:
+            self.function.gradient(x, out=out)
+            out *= self.scalar  
+#        except NotImplementedError:
+#            print("{} is not differentiable".format(type(self.function).__name__))                         
 
+    def proximal(self, x, tau, out=None):
+        
+        r"""Returns the proximal operator of the scaled function.
+        
+        .. math:: \mathrm{prox}_{\tau G}(x) = \mathrm{prox}_{(\tau\alpha) F}(x)
+        
+        """        
+#        if out is None:
+        return self.function.proximal(x, tau*self.scalar, out=out)     
+#        else:
+#            self.function.proximal(x, tau*self.scalar, out = out)
+
+    def proximal_conjugate(self, x, tau, out = None):
+        r"""This returns the proximal operator for the function at x, tau
+        """
+        if out is None:
+            return self.scalar * self.function.proximal_conjugate(x/self.scalar, tau/self.scalar)
+        else:
+            self.function.proximal_conjugate(x/self.scalar, tau/self.scalar, out=out)
+            out *= self.scalar
+
+    def function(self):
```

Comment:
The member `function` is already accessible. Why the method `function`?
This has to be removed.